File: chesslib/fen.cpp
#include "fen.h"

#include "CastleOptions.h"
#include "PieceOnBoard.h"
#include "position_factory.h"
#include "require.h"
#include "string_utils.h"

#include <cctype>

namespace {
    constexpr char ROWS_SEPARATOR = '/';
    constexpr char FEN_WHITE_TO_MOVE = 'w';
    constexpr char FEN_BLACK_TO_MOVE = 'b';
    constexpr char FEN_NO_ONE_CAN_CASTLE = '-';
    constexpr char FEN_WHITE_CAN_CASTLE_SHORT = 'K';
    constexpr char FEN_WHITE_CAN_CASTLE_LONG = 'Q';
    constexpr char FEN_BLACK_CAN_CASTLE_SHORT = 'k';
    constexpr char FEN_BLACK_CAN_CASTLE_LONG = 'q';
    constexpr char FEN_CANNOT_CAPTURE_EN_PASSANT = '-';
}

namespace chesslib {

namespace {

player_t decodePlayer(char c)
{
    return std::islower(c) ? Black : White;
}

piece_type_t decodePieceType(char c)
{
    switch (std::tolower(c)) {
    case 'p':
        return Pawn;
    case 'n':
        return Knight;
    case 'b':
        return Bishop;
    case 'r':
        return Rook;
    case 'q':
        return Queen;
    case 'k':
        return King;
    default:
        FAIL();
        return 0;
    }
}

char encodePlayerPieceType(player_t player, piece_type_t pieceType)
{
    char pieceChar;
    switch (pieceType)
    {
    case Pawn:
        pieceChar = 'p';
        break;
    case Knight:
        pieceChar = 'n';
        break;
    case Bishop:
        pieceChar = 'b';
        break;
    case Rook:
        pieceChar = 'r';
        break;
    case Queen:
        pieceChar = 'q';
        break;
    case King:
        pieceChar = 'k';
        break;
    default:
        FAIL();
        pieceChar = '\0';
    }
    return player == Black ? pieceChar : std::toupper(pieceChar);
}

std::vector<PieceOnBoard> decodeBoard(const std::string& boardString)
{
    std::vector<PieceOnBoard> decodedPieces;

    const auto rowStrings = split(boardString, ROWS_SEPARATOR);
    REQUIRE(rowStrings.size() == ROW_COUNT);

    square_t currentRow = ROW_COUNT;
    for (auto& rowString : rowStrings) {
        --currentRow;
        square_t currentColumn = 0;
        for (auto c : rowString) {
            REQUIRE(currentColumn < COLUMN_COUNT);
            if (std::isdigit(c)) {
                currentColumn += c - '0';
            }
            else {
                decodedPieces.push_back({ decodePlayer(c), decodePieceType(c),
                    encodeSquare(currentRow, currentColumn) });
                ++currentColumn;
            }
        }
        REQUIRE(currentColumn == COLUMN_COUNT);
    }
    REQUIRE(currentRow == 0);

    return decodedPieces;
}

std::string encodeBoard(const Board& board)
{
    std::string encodedBoard;

    for (square_t currentRow = MAX_ROW; currentRow >= 0; --currentRow) {
        std::string encodedRow;
        fastint emptySquaresCounter = 0;
        for (square_t currentColumn = 0; currentColumn < COLUMN_COUNT; ++currentColumn) {
            const square_t currentSquare = encodeSquare(currentRow, currentColumn);
            if (board.isEmpty(currentSquare)) {
                ++emptySquaresCounter;
            }
            else {
                if (emptySquaresCounter != 0) {
                    encodedRow += std::to_string(emptySquaresCounter);
                    emptySquaresCounter = 0;
                }
                encodedRow.push_back(encodePlayerPieceType(
                    board.getPlayer(currentSquare),
                    board.getPieceTypeAt(currentSquare)));
            }
        }
        if (emptySquaresCounter != 0) {
            encodedRow += std::to_string(emptySquaresCounter);
        }
        encodedBoard += encodedRow;
        if (currentRow != 0) {
            encodedBoard.push_back(ROWS_SEPARATOR);
        }
    }

    return encodedBoard;
}

player_t decodePlayerToMove(const std::string& encoded)
{
    REQUIRE(encoded.size() == 1);
    switch (std::tolower(encoded.front())) {
    case FEN_BLACK_TO_MOVE:
        return Black;
    case FEN_WHITE_TO_MOVE:
        return White;
    default:
        FAIL();
        return 0;
    }
}

std::string encodePlayerToMove(player_t playerToMove)
{
    return { playerToMove == Black ? FEN_BLACK_TO_MOVE : FEN_WHITE_TO_MOVE };
}

std::array<CastleOptions, PLAYER_COUNT> decodeCastleOptions(const std::string& encoded)
{
    std::array<CastleOptions, PLAYER_COUNT> result;
    
    result[Black].setCanCastleShort(contains(encoded, FEN_BLACK_CAN_CASTLE_SHORT));
    result[Black].setCanCastleLong(contains(encoded, FEN_BLACK_CAN_CASTLE_LONG));
    result[White].setCanCastleShort(contains(encoded, FEN_WHITE_CAN_CASTLE_SHORT));
    result[White].setCanCastleLong(contains(encoded, FEN_WHITE_CAN_CASTLE_LONG));

    return result;
}

std::string encodeCastleOptions(const std::array<CastleOptions, PLAYER_COUNT>& castleOptions)
{
    std::string encodedCastleOptions;

    if (castleOptions[White].isCanCastleShort()) {
        encodedCastleOptions.push_back(FEN_WHITE_CAN_CASTLE_SHORT);
    }
    if (castleOptions[White].isCanCastleLong()) {
        encodedCastleOptions.push_back(FEN_WHITE_CAN_CASTLE_LONG);
    }
    if (castleOptions[Black].isCanCastleShort()) {
        encodedCastleOptions.push_back(FEN_BLACK_CAN_CASTLE_SHORT);
    }
    if (castleOptions[Black].isCanCastleLong()) {
        encodedCastleOptions.push_back(FEN_BLACK_CAN_CASTLE_LONG);
    }
    if (encodedCastleOptions.empty()) {
        encodedCastleOptions = FEN_NO_ONE_CAN_CASTLE;
    }

    return encodedCastleOptions;
}

OptionalColumn decodeEnPassantColumn(const std::string& encoded)
{
    if (encoded.size() > 1) {
        const square_t result = encoded.front() - 'a';
        REQUIRE(result >= 0 && result < COLUMN_COUNT);
        return OptionalColumn::fromColumn(result);
    }
    else {
        return OptionalColumn();
    }
}

std::string encodeEnPassantColumn(player_t playerToMove, OptionalColumn enPassantColumn)
{
    std::string encodedEnPassantColumn;

    if (enPassantColumn.hasValue()) {
        encodedEnPassantColumn.push_back('a' + enPassantColumn.getColumn());
        encodedEnPassantColumn.push_back(playerToMove == Black ? '3' : '6');
    }
    else {
        encodedEnPassantColumn = FEN_CANNOT_CAPTURE_EN_PASSANT;
    }

    return encodedEnPassantColumn;
}

uint32_t decodeHalfmoveClock(const std::string& encoded)
{
    return static_cast<uint32_t>(std::stoi(encoded));
}

std::string encodeHalfmoveClock()
{
    // TODO: Return actual value
    return std::to_string(0);
}

uint32_t decodeFullmoveNumber(const std::string& encoded)
{
    return static_cast<uint32_t>(std::stoi(encoded));
}

std::string encodeFullmoveNumber()
{
    // TODO: Return actual value
    return std::to_string(1);
}

}  // namespace

Position decodeFen(const std::string& fenString)
{
    const auto tokens = split(fenString, ' ');
    REQUIRE(tokens.size() >= 4);

    auto pieces = decodeBoard(tokens[0]);
    const player_t playerToMove = decodePlayerToMove(tokens[1]);
    auto castleOptions = decodeCastleOptions(tokens[2]);
    const auto enPassantColumn = decodeEnPassantColumn(tokens[3]);

    auto halfmoveClock = tokens.size() > 4 ? decodeHalfmoveClock(tokens[4]) : 0;
    auto fullmoveNumber = tokens.size() > 5 ? decodeFullmoveNumber(tokens[5]) : 0;

    return createPosition(
        std::move(pieces),
        playerToMove,
        enPassantColumn,
        castleOptions[White],
        castleOptions[Black],
        halfmoveClock,
        fullmoveNumber);
}

std::string encodeFen(const Position& position)
{
    return join({
            encodeBoard(position.getBoard()),
            encodePlayerToMove(position.getPlayerToMove()),
            encodeCastleOptions({ position.getCastleOptions(Black), position.getCastleOptions(White) }),
            encodeEnPassantColumn(position.getPlayerToMove(), position.getEnPassantColumn()),
            encodeHalfmoveClock(),
            encodeFullmoveNumber()
        },
        ' ');
}

}  // namespace chesslib

File: android/chesslibkt/src/main/java/ru/maksimbulva/chesslibkt/Fen.kt
package ru.maksimbulva.chesslibkt

import ru.maksimbulva.chesslibkt.board.Board
import ru.maksimbulva.chesslibkt.board.PieceOnBoard
import ru.maksimbulva.chesslibkt.board.Square
import ru.maksimbulva.chesslibkt.position.CastleOptions
import ru.maksimbulva.chesslibkt.position.Position
import ru.maksimbulva.chesslibkt.position.PositionFactory
import ru.maksimbulva.chesslibkt.position.PositionMoveCounters

object Fen {
    private const val ROWS_SEPARATOR = '/'
    private const val FEN_WHITE_TO_MOVE = 'w'
    private const val FEN_BLACK_TO_MOVE = 'b'
    private const val FEN_NO_ONE_CAN_CASTLE = '-'
    private const val FEN_WHITE_CAN_CASTLE_SHORT = 'K'
    private const val FEN_WHITE_CAN_CASTLE_LONG = 'Q'
    private const val FEN_BLACK_CAN_CASTLE_SHORT = 'k'
    private const val FEN_BLACK_CAN_CASTLE_LONG = 'q'
    private const val FEN_CANNOT_CAPTURE_EN_PASSANT = '-'

    private val pieceToChar = mapOf(
        Piece.Pawn to 'p',
        Piece.Knight to 'n',
        Piece.Bishop to 'b',
        Piece.Rook to 'r',
        Piece.Queen to 'q',
        Piece.King to 'k'
    )

    fun decodeFen(fenString: String): Position {
        val tokens = fenString.split(' ')
        return PositionFactory.createPosition(
            pieces = decodeBoard(tokens[0]),
            playerToMove = decodePlayerToMove(tokens[1]),
            castleOptions = decodeCastleOptions(tokens[2]),
            enPassantColumn = decodeEnPassantColumn(tokens[3]),
            halfmoveClock = tokens.takeIf { tokens.size > 4 }
                .run { this?.let { decodeHalfmoveClock(it[4]) } ?: 0 },
            fullmoveNumber = tokens.takeIf { tokens.size > 5 }
                .run { this?.let { decodeFullmoveNumber(it[5]) } ?: 0 }
        )
    }

    fun encodeFen(position: Position): String {
        return listOf(
            encodeBoard(position.board),
            encodePlayerToMove(position.playerToMove),
            encodeCastleOptions(
                mapOf(
                    Player.Black to position.getCastleOptions(Player.Black),
                    Player.White to position.getCastleOptions(Player.White)
                )
            ),
            encodeEnPassantColumn(position.playerToMove, position.enPassantColumn),
            encodeHalfmoveClock(position.halfmoveClock),
            encodeFullmoveNumber(position.fullmoveNumber)
        ).joinToString(separator = "")
    }

    private fun decodeBoard(boardString: String): Collection<PieceOnBoard> {
        val decodedPieces = mutableListOf<PieceOnBoard>()

        val rowStrings = boardString.split(ROWS_SEPARATOR)
        require(rowStrings.size == Board.ROW_COUNT)

        Board.RowsRange.reversed().forEachIndexed { index, currentRow ->
            var currentColumn = 0
            for (c in rowStrings[index]) {
                require(currentColumn < Board.COLUMN_COUNT)
                currentColumn += when {
                    c.isDigit() -> c - '0'
                    else -> {
                        decodedPieces.add(
                            PieceOnBoard(
                                decodePlayer(c),
                                decodePieceType(c),
                                Square(currentRow, currentColumn)
                            )
                        )
                        1
                    }
                }
            }
            require(currentColumn == Board.COLUMN_MAX + 1)
        }

        return decodedPieces
    }

    private fun encodeBoard(board: Board): String
    {
        TODO()
        /*std::string encodedBoard;

        for (square_t currentRow = MAX_ROW; currentRow >= 0; --currentRow) {
        std::string encodedRow;
        fastint emptySquaresCounter = 0;
        for (square_t currentColumn = 0; currentColumn < COLUMN_COUNT; ++currentColumn) {
        const square_t currentSquare = encodeSquare(currentRow, currentColumn);
        if (board.isEmpty(currentSquare)) {
            ++emptySquaresCounter;
        }
        else {
            if (emptySquaresCounter != 0) {
                encodedRow += std::to_string(emptySquaresCounter);
                emptySquaresCounter = 0;
            }
            encodedRow.push_back(encodePlayerPieceType(
                board.getPlayer(currentSquare),
                board.getPieceTypeAt(currentSquare)));
        }
    }
        if (emptySquaresCounter != 0) {
            encodedRow += std::to_string(emptySquaresCounter);
        }
        encodedBoard += encodedRow;
        if (currentRow != 0) {
            encodedBoard.push_back(ROWS_SEPARATOR);
        }
    }

        return encodedBoard;*/
    }

    private fun decodePlayer(c: Char): Player {
        return if (c.isLowerCase()) {
            Player.Black
        } else {
            Player.White
        }
    }

    private fun decodePieceType(c: Char): Piece {
        val pieceCharLower = c.toLowerCase()
        return pieceToChar.filter { it.value == pieceCharLower }.map { it.key }.first()
    }

    private fun encodePlayerPieceType(player: Player, piece: Piece): Char {
        return pieceToChar.getValue(piece).run {
            when (player) {
                Player.Black -> this
                Player.White -> toUpperCase()
            }
        }
    }

    private fun decodePlayerToMove(encoded: String): Player {
        require(encoded.length == 1)
        return when (encoded.first().toLowerCase()) {
            FEN_BLACK_TO_MOVE -> Player.Black
            FEN_WHITE_TO_MOVE -> Player.White
            else -> throw ChesslibException()
        }
    }

    private fun encodePlayerToMove(playerToMove: Player): Char {
        return when (playerToMove) {
            Player.Black -> FEN_BLACK_TO_MOVE
            Player.White -> FEN_WHITE_TO_MOVE
        }
    }

    private fun decodeCastleOptions(encoded: String): Map<Player, CastleOptions> {
        val blackCastleOptions = CastleOptions.None
            .run {
                if (FEN_BLACK_CAN_CASTLE_SHORT in encoded) {
                    setCanCastleShort(true)
                } else {
                    this
                }
            }
            .run {
                if (FEN_BLACK_CAN_CASTLE_LONG in encoded) {
                    setCanCastleLong(true)
                } else {
                    this
                }
            }

        val whiteCastleOptions = CastleOptions.None
            .run {
                if (FEN_WHITE_CAN_CASTLE_SHORT in encoded) {
                    setCanCastleShort(true)
                } else {
                    this
                }
            }
            .run {
                if (FEN_WHITE_CAN_CASTLE_LONG in encoded) {
                    setCanCastleLong(true)
                } else {
                    this
                }
            }

        return mapOf(
            Player.Black to blackCastleOptions,
            Player.White to whiteCastleOptions
        )
    }

    private fun encodeCastleOptions(castleOptionsByPlayer: Map<Player, CastleOptions>): String {
        return with (castleOptionsByPlayer) {
            listOfNotNull(
                FEN_BLACK_CAN_CASTLE_SHORT.takeIf { getValue(Player.Black).isCanCastleShort },
                FEN_BLACK_CAN_CASTLE_LONG.takeIf { getValue(Player.Black).isCanCastleLong },
                FEN_WHITE_CAN_CASTLE_SHORT.takeIf { getValue(Player.White).isCanCastleShort },
                FEN_WHITE_CAN_CASTLE_LONG.takeIf { getValue(Player.White).isCanCastleLong }
            ).joinToString(separator = "").run {
                if (isEmpty()) FEN_NO_ONE_CAN_CASTLE.toString() else this
            }
        }
    }

    private fun decodeEnPassantColumn(encoded: String): Int? {
        return if (encoded.length > 1) {
            (encoded.first().toLowerCase() - 'a').also {
                check(it in 0..Board.COLUMN_MAX)
            }
        } else {
            null
        }
    }

    private fun encodeEnPassantColumn(playerToMove: Player, enPassantColumn: Int?): String {
        return if (enPassantColumn != null) {
            val columnChar = 'a' + enPassantColumn
            val rowChar = when (playerToMove) {
                Player.Black -> '3'
                Player.White -> '6'
            }
            "$columnChar$rowChar"
        } else {
            FEN_CANNOT_CAPTURE_EN_PASSANT.toString()
        }
    }

    private fun decodeHalfmoveClock(encoded: String): Int {
        return encoded.toInt()
    }

    private fun encodeHalfmoveClock(halfmoveClock: Int): String {
        return halfmoveClock.toString()
    }

    private fun decodeFullmoveNumber(encoded: String): Int {
        return encoded.toInt()
    }

    private fun encodeFullmoveNumber(fullmoveNumber: Int): String {
        return fullmoveNumber.toString()
    }
}
